src/par_iter.rs

```diff
@@ -0,0 +1,145 @@
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+//! Connecting traits and types to bridge `Iterator` and `ParallelIterator`.
+
+use std::sync::Arc;
+use std::sync::atomic::{AtomicUsize, AtomicBool, Ordering};
+
+use crossbeam_deque::{Deque, Stealer, Steal};
+use rayon::prelude::*;
+use rayon::iter::plumbing::*;
+use rayon::{scope, current_num_threads};
+
+/// Conversion trait to convert an `Iterator` to a `ParallelIterator`.
+///
+/// This needs to be distinct from `IntoParallelIterator` because that trait is already implemented
+/// on a few `Iterator`s, like `std::ops::Range`.
+pub trait AsParallel {
+    /// What is the type of the output `ParallelIterator`?
+    type Iter: ParallelIterator<Item = Self::Item>;
+
+    /// What is the `Item` of the output `ParallelIterator`?
+    type Item: Send;
+
+    /// Convert this type to a `ParallelIterator`.
+    fn as_parallel(self) -> Self::Iter;
+}
+
+impl<T: IntoIterator + Send> AsParallel for T
+    where T::Item: Send
+{
+    type Iter = IterParallel<T>;
+    type Item = T::Item;
+
+    fn as_parallel(self) -> Self::Iter {
+        IterParallel {
+            iter: self,
+        }
+    }
+}
+
+pub struct IterParallel<Iter> {
+    iter: Iter,
+}
+
+impl<Iter: IntoIterator + Send> ParallelIterator for IterParallel<Iter>
+    where Iter::Item: Send
+{
+    type Item = Iter::Item;
+
+    fn drive_unindexed<C>(self, consumer: C) -> C::Result
+        where C: UnindexedConsumer<Self::Item>
+    {
+        scope(|s| {
+            let done = Arc::new(AtomicBool::new(false));
+            let deque = Deque::new();
+            let stealer = deque.stealer();
+
+            let signal = done.clone();
+            s.spawn(move |_| {
+                let mut iter = self.iter.into_iter();
+
+                loop {
+                    if !signal.load(Ordering::SeqCst) { break; }
+
+                    match iter.next() {
+                        Some(it) => deque.push(it),
+                        _ => break,
+                    }
+                }
+
+                // if we got here, either the iterator is empty or the consumer is full - if the
+                // former, let's signal to the consumer to stop allowing splits
+                signal.store(true, Ordering::SeqCst);
+            });
+
+            let result = bridge_unindexed(IterParallelProducer {
+                split_count: Arc::new(AtomicUsize::new(current_num_threads())),
+                done: done.clone(),
+                items: stealer,
+            }, consumer);
+
+            // if we're here, either the iterator is empty or the consumer is full - if the latter,
+            // let's signal back to the iterator to stop
+            done.store(true, Ordering::SeqCst);
+
+            result
+        })
+    }
+}
+
+struct IterParallelProducer<T> {
+    split_count: Arc<AtomicUsize>,
+    done: Arc<AtomicBool>,
```

Comment:
I think simple references will suffice, rather than `Arc`.  The producer doesn't need to own the values, just borrow them from the local scope of `drive_unindexed`.